### DISCUSSION TITLE
Start time handling

### DISCUSSION THREAD

**Discussion by mundschenk-at** [astrandb/miele#37] [General]
Would it be more useful to give the `*_start_time` as an absolute time or is that something better left to template sensors? Anyway, I think `[0, 0]` should be treated as `none`.

**Comment by astrandb** [comment]
As a general principle, I think we should avoid to convert or interpret values from the API. A template sensor would solve this.
[0, 0] i.e. 0 minutes is a time offset, and It would complicate for automations and tempates to handle None also.

**Comment by mundschenk-at** [comment]
Yeah, but `[0, 0]` for `startTime` doesn't mean "now", it means "there is no start time set".

**Reply by astrandb** [reply]
I understand. But how is the delayed start handled by the API? Does it start the countdown as soon as you set the startTime to a non-zero value or do you have to send start somewhere? How is it handled in the app?
I have only pre-smart washing machines and tumble dryers from Miele and there you press a delay button a number of times and then press start button.
If we set the internal value to `none` it will show `Unknown` in frontend. Is that better than `0`?

**Reply by mundschenk-at** [reply]
For my dishwasher, you have to set the time - either by entering a begin time or an end time where it will calculate the begin time for you - and then press "OK" twice, once to acknowledge the time value and once to actually set the program. I'll have to check when the API updates its value. For the washer-dryer, I haven't yet used the feature. I'm assuming it works the same way, but via the touch screen instead of physical buttons.

In any case, on the device you don't actually ever see the offset, just the resulting timestamp (begin or end time). In the app, I'll have to double check.

**Reply by astrandb** [reply]
Check if startTime is an offset or an absolute timestamp in the API. Can be different meaning in status and actions PUT command.

**Reply by mundschenk-at** [reply]
It is an offset in the API, I checked that yesterday. (Probably for compatibility with older devices that didn't have the notion of wall time/time synchronization.)

**Reply by mundschenk-at** [reply]
> For my dishwasher, you have to set the time - either by entering a begin time or an end time where it will calculate the begin time for you - and then press "OK" twice, once to acknowledge the time value and once to actually set the program. I'll have to check when the API updates its value.

I've got to correct myself there: The two "OK" presses are needed to set the time and once it is set, the dishwasher is fully programmed, you can only start over by switching it off and on again. The API sets the `startTime` offset on the second button press. Oh, and there are three choices: "Start at", "End at", "SmartStart". The latter starts so that the dishes are ready by the set time, or earlier if the energy company sends a signal that a cheaper rate is now available. I don't know which (if at all) energy providers do that. Maybe it is based on some standard, but I haven't been able to find any more detailed information on this feature.

**Comment by astrandb** [comment]
I still think we should prefer to use native data and avoid calculation/conversion in the integration.
But I can be reasonable now and then. :wink:
What if we keep the relevant sensors and services as they are today and add a set of sensors and services that handles absolute time -  "Start at" only.

**Comment by derbruns** [comment]
I hope somebody can help me. I would like to program the dishwasher so that it does not start directly but at a specified time, e.g. 10:30 p.m.
Can someone give me a specific example?
I've tried different things but can't get it to work
Examplei have tried

`  - service: miele.generic_action
    data:
      sensor.miele_g7160_start_time: [22, 30]
    target:
      device_id: xyz123
      entity_id: sensor.miele_g7160_start_time`

switched on the dishwasher and selected the program are working.

**Reply by mundschenk-at** [reply]
The sensor entities are not something you can use in the action service as the Miele API has no knowledge of them. If selecting a program and starting the dishwasher via the action service works, I'd suggest doing the scheduling in a HA automation and only calling the action service when it's time.

(Also, this discussion is not a general help topic. For follow-ups, you should post a separate question [here](https://github.com/astrandb/miele/discussions) or maybe better yet, on the HA Community forum.)
